src/Compiler/Facilities/DiagnosticsLogger.fs

```diff
@@ -905,16 +894,71 @@ type StackGuard(maxDepth: int, name: string) =
+    let mutable errorCount = 0
+
+    let computationsWithLoggers, diagnosticsReady =
+        [
+            for i, computation in computations |> Seq.indexed do
+                let diagnosticsReady =
+                    TaskCompletionSource<_>(TaskCreationOptions.RunContinuationsAsynchronously)
+
+                // Diagnostics logger utilizing the common error count.
+                let logger =
+                    { new CapturingDiagnosticsLogger($"CaptureDiagnosticsConcurrently {i}", ?eagerFormat = eagerFormat) with
+                        override _.DiagnosticSink(d, severity) =
+                            base.DiagnosticSink(d, severity)
+
+                            if severity = FSharpDiagnosticSeverity.Error then
+                                Interlocked.Increment &errorCount |> ignore
+
+                        override _.ErrorCount = errorCount
+                    }
+
+                // Inject capturing loger into the computation. Signal the TaskCompletionSource when done.
+                let computationsWithLoggers =
+                    async {
+                        SetThreadDiagnosticsLoggerNoUnwind logger
+
+                        try
+                            return! computation
+                        finally
+                            diagnosticsReady.SetResult logger
+                    }
+
+                computationsWithLoggers, diagnosticsReady
+        ]
+        |> List.unzip
+
+    // Commit diagnostics from computations as soon as it is possible, preserving the order.
+    let replayDiagnostics =
+        backgroundTask {
+            let target = DiagnosticsThreadStatics.DiagnosticsLogger
+
+            for tcs in diagnosticsReady do
+                let! finishedLogger = tcs.Task
+                finishedLogger.CommitDelayedDiagnostics target
+
+            return target
+        }
 
-    member _.GetLoggerForTask(name) : DiagnosticsLogger =
-        let logger = CapturingDiagnosticsLogger(name)
-        loggers.Add logger
-        logger
+    member val Computations = computationsWithLoggers |> Seq.ofList
+    member val ReplayDiagnostics = replayDiagnostics
 
-    interface IDisposable with
-        member _.Dispose() =
-            for logger in loggers do
-                logger.CommitDelayedDiagnostics target
+module MultipleDiagnosticsLoggers =
+    // Capture diagnostics from multiple computations.
+    let run method computations =
+        let forks = CaptureDiagnosticsConcurrently(computations)
+
+        async {
+            try
+                // We want to restore the current diagnostics context when finished.
+                use _ = new CompilationGlobalsScope()
+                return! forks.Computations |> method
+            finally
+                forks.ReplayDiagnostics.Wait()
+        }
+
+    let Parallel computations = computations |> run Async.Parallel
+    let Sequential computations = computations |> run Async.Sequential
```

Comment:
Async.Sequential still uses thread pool, IIUC. Do we want a custom implementation instead?

Reply:
It would make sense for `Sequential` to keep things running on the same thread, without using the ThreadPool.

Reply:
Yes, or at least start on the same thread.